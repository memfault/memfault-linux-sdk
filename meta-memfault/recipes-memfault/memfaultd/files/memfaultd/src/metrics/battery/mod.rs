//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
