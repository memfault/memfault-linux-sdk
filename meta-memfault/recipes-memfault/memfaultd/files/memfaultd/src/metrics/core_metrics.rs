//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";
