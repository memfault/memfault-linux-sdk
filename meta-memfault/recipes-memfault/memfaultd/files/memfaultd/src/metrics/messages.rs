//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::path::PathBuf;

use eyre::Result;
use ssf::{Message, MsgMailbox};

use crate::network::NetworkConfig;

use super::{KeyedMetricReading, SessionName};

/// Allows KeyedMetricReading to be sent as a message. The `ssf` framework will
/// automatically support sending `Vec<KeyedMetricReading>` as well.
impl Message for KeyedMetricReading {
    type Reply = Result<()>;
}

/// Syntactic Sugar because this is the most-used type of mailbox in memfaultd.
pub type MetricsMBox = MsgMailbox<Vec<KeyedMetricReading>>;

pub enum SessionEventMessage {
    StartSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
    },
    StopSession {
        name: SessionName,
        readings: Vec<KeyedMetricReading>,
        mar_staging_area: PathBuf,
        network_config: NetworkConfig,
    },
}

impl Message for SessionEventMessage {
    type Reply = Result<()>;
}
