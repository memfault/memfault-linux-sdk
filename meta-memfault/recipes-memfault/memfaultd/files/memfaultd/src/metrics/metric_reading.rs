//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::str::FromStr;

use chrono::{Duration, Utc};
use eyre::{eyre, ErrReport};
use serde::{Deserialize, Serialize};

use super::{MetricStringKey, MetricTimestamp};

use crate::util::serialization::float_to_duration;

#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub enum MetricReading {
    /// TimeWeightedAverage readings will be aggregated based on
    /// the time the reading was captured over.
    TimeWeightedAverage {
        value: f64,
        timestamp: MetricTimestamp,
        /// Time period considered for this reading. This is only used to give a "time-weight" to the first
        /// value in the series (when using time-weighted averages). For future values we will use the time
        /// difference we measure between the two points         
        /// In doubt, it's safe to use Duration::from_secs(0) here. This means the first value will be ignored.
        #[serde(with = "float_to_duration")]
        interval: Duration,
    },
    /// A non-decreasing monotonic sum. Within a metric report, Counter readings are summed together.
    Counter {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Gauges are absolute values. We keep the latest value collected during a metric report.
    Gauge {
        value: f64,
        timestamp: MetricTimestamp,
    },
    /// Histogram readings are averaged together by dividing the sum of the values
    /// by the number of readings over the duration of a metric report.
    Histogram {
        value: f64,
        timestamp: MetricTimestamp,
    },
}

#[derive(Debug, Serialize, Clone, Deserialize)]
pub struct KeyedMetricReading {
    pub name: MetricStringKey,
    pub value: MetricReading,
}

impl KeyedMetricReading {
    pub fn new(name: MetricStringKey, value: MetricReading) -> Self {
        Self { name, value }
    }
}

impl FromStr for KeyedMetricReading {
    type Err = ErrReport;

    /// Deserialize a string in the form <MetricStringKey>=<f64> to a Gauge metric reading
    ///
    /// Currently deserialization to a KeyedMetricReading with any other type of MetricReading
    /// as its value is not supported
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (key, value_str) = s.split_once('=').ok_or(eyre!(
            "Gauge metric reading should be specified as KEY=VALUE"
        ))?;

        // Let's ensure the key is valid first:
        let metric_key = MetricStringKey::from_str(key).map_err(|e| eyre!(e))?;
        let value =
            f64::from_str(value_str).map_err(|e| eyre!("Invalid value {}: {}", value_str, e))?;

        let reading = MetricReading::Gauge {
            value,
            timestamp: Utc::now(),
        };
        Ok(KeyedMetricReading::new(metric_key, reading))
    }
}
