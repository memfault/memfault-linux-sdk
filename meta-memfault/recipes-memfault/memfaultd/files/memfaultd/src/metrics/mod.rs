//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
