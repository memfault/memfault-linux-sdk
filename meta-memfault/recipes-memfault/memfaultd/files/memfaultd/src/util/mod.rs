//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod disk_backed;
pub use disk_backed::*;
mod die;
pub mod disk_size;
pub mod patterns;
pub use die::*;
pub mod can_connect;
pub mod circular_queue;
pub mod fs;
pub mod io;
pub mod ipc;
pub mod math;
pub mod mem;
pub mod output_arg;
pub mod path;
pub mod persistent_rate_limiter;
pub mod pid_file;
#[cfg(feature = "logging")]
pub mod rate_limiter;
pub mod serialization;
pub mod string;
pub mod system;
pub mod task;
#[cfg(feature = "logging")]
pub mod tcp_server;
pub mod time_measure;
pub mod zip;
