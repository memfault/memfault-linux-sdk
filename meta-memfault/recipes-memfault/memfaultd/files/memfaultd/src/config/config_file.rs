//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use eyre::{eyre, Context};
use serde::{Deserialize, Serialize};
use std::net::IpAddr;
use std::time::Duration;
use std::{
    collections::{HashMap, HashSet},
    num::NonZeroU32,
};
use std::{net::SocketAddr, path::PathBuf};

use crate::metrics::{MetricStringKey, SessionName};
use crate::util::*;
use crate::util::{path::AbsolutePath, serialization::*};

#[derive(Serialize, Deserialize, Debug)]
pub struct MemfaultdConfig {
    pub persist_dir: AbsolutePath,
    pub tmp_dir: Option<AbsolutePath>,
    #[serde(rename = "tmp_dir_min_headroom_kib", with = "kib_to_usize")]
    pub tmp_dir_min_headroom: usize,
    pub tmp_dir_min_inodes: usize,
    #[serde(rename = "tmp_dir_max_usage_kib", with = "kib_to_usize")]
    pub tmp_dir_max_usage: usize,
    #[serde(rename = "upload_interval_seconds", with = "seconds_to_duration")]
    pub upload_interval: Duration,
    #[serde(rename = "heartbeat_interval_seconds", with = "seconds_to_duration")]
    pub heartbeat_interval: Duration,
    pub enable_data_collection: bool,
    pub enable_dev_mode: bool,
    pub software_version: Option<String>,
    pub software_type: Option<String>,
    pub project_key: String,
    pub base_url: String,
    pub swupdate: SwUpdateConfig,
    pub reboot: RebootConfig,
    pub coredump: CoredumpConfig,
    #[serde(rename = "fluent-bit")]
    pub fluent_bit: FluentBitConfig,
    pub logs: LogsConfig,
    pub mar: MarConfig,
    pub http_server: HttpServerConfig,
    pub battery_monitor: Option<BatteryMonitorConfig>,
    pub connectivity_monitor: Option<ConnectivityMonitorConfig>,
    pub sessions: Option<Vec<SessionConfig>>,
    pub metrics: MetricReportConfig,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct SwUpdateConfig {
    pub input_file: PathBuf,
    pub output_file: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct RebootConfig {
    pub last_reboot_reason_file: PathBuf,
    pub capture_pstore: bool,
}

#[derive(Serialize, Deserialize, Debug, Copy, Clone)]
pub enum CoredumpCompression {
    #[serde(rename = "gzip")]
    Gzip,
    #[serde(rename = "none")]
    None,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
#[serde(tag = "type")]
pub enum CoredumpCaptureStrategy {
    #[serde(rename = "threads")]
    /// Only capture the stacks of the threads that were running at the time of the crash.
    Threads {
        #[serde(rename = "max_thread_size_kib", with = "kib_to_usize")]
        max_thread_size: usize,
    },
    #[serde(rename = "kernel_selection")]
    /// Keep in the coredump what the kernel selected to be included in the coredump.
    /// See https://man7.org/linux/man-pages/man5/core.5.html for more details.
    KernelSelection,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct CoredumpConfig {
    pub compression: CoredumpCompression,
    #[serde(rename = "coredump_max_size_kib", with = "kib_to_usize")]
    pub coredump_max_size: usize,
    pub rate_limit_count: u32,
    #[serde(rename = "rate_limit_duration_seconds", with = "seconds_to_duration")]
    pub rate_limit_duration: Duration,
    pub capture_strategy: CoredumpCaptureStrategy,
    pub log_lines: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct FluentBitConfig {
    pub extra_fluentd_attributes: Vec<String>,
    pub bind_address: SocketAddr,
    pub max_buffered_lines: usize,
    pub max_connections: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct HttpServerConfig {
    pub bind_address: SocketAddr,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
pub enum LogSource {
    #[serde(rename = "fluent-bit")]
    FluentBit,
    #[cfg(feature = "systemd")]
    #[serde(rename = "journald")]
    Journald,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LogsConfig {
    #[serde(rename = "rotate_size_kib", with = "kib_to_usize")]
    pub rotate_size: usize,

    #[serde(rename = "rotate_after_seconds", with = "seconds_to_duration")]
    pub rotate_after: Duration,

    #[serde(with = "number_to_compression")]
    pub compression_level: Compression,

    pub max_lines_per_minute: NonZeroU32,

    pub log_to_metrics: Option<LogToMetricsConfig>,

    pub storage: StorageConfig,

    pub source: LogSource,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
pub enum StorageConfig {
    #[serde(rename = "disabled")]
    Disabled,
    #[serde(rename = "persist")]
    Persist,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LogToMetricsConfig {
    pub rules: Vec<LogToMetricRule>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "type")]
pub enum LogToMetricRule {
    #[serde(rename = "count_matching")]
    CountMatching {
        /// Regex applied on the MESSAGE field
        pattern: String,
        metric_name: String,
        /// List of key-value that must exactly match before the regexp is applied
        #[serde(default)]
        filter: HashMap<String, String>,
    },
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MarConfig {
    #[serde(rename = "mar_file_max_size_kib", with = "kib_to_usize")]
    pub mar_file_max_size: usize,
    #[serde(rename = "mar_entry_max_age_seconds", with = "seconds_to_duration")]
    pub mar_entry_max_age: Duration,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct BatteryMonitorConfig {
    pub battery_info_command: String,
    #[serde(with = "seconds_to_duration")]
    pub interval_seconds: Duration,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ConnectivityMonitorConfig {
    #[serde(with = "seconds_to_duration")]
    pub interval_seconds: Duration,
    pub targets: Vec<ConnectivityMonitorTarget>,
    #[serde(
        with = "seconds_to_duration",
        default = "default_connection_check_timeout"
    )]
    pub timeout_seconds: Duration,
}
fn default_connection_check_timeout() -> Duration {
    Duration::from_secs(10)
}

#[derive(Serialize, Deserialize, Debug)]
pub struct MetricReportConfig {
    pub enable_daily_heartbeats: bool,
    pub system_metric_collection: SystemMetricConfig,
    pub statsd_server: Option<StatsDServerConfig>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ConnectivityMonitorTarget {
    #[serde(default = "default_connection_check_protocol")]
    pub protocol: ConnectionCheckProtocol,
    pub host: IpAddr,
    pub port: u16,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum ConnectionCheckProtocol {
    Tcp,
}

fn default_connection_check_protocol() -> ConnectionCheckProtocol {
    ConnectionCheckProtocol::Tcp
}

#[derive(Serialize, Clone, Deserialize, Debug)]
pub struct SessionConfig {
    pub name: SessionName,
    pub captured_metrics: Vec<MetricStringKey>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct StatsDServerConfig {
    pub bind_address: SocketAddr,
}

#[derive(Clone, Serialize, Deserialize, Debug)]
pub struct SystemMetricConfig {
    pub enable: bool,
    #[serde(with = "seconds_to_duration")]
    pub poll_interval_seconds: Duration,
    pub processes: Option<HashSet<String>>,
    pub disk_space: Option<HashSet<String>>,
    pub network_interfaces: Option<HashSet<String>>,
}

use flate2::Compression;
use serde_json::Value;
use std::fs;
use std::path::Path;

use crate::config::utils::{software_type_is_valid, software_version_is_valid};

pub struct JsonConfigs {
    /// Built-in configuration and System configuration
    pub base: Value,
    /// Runtime configuration
    pub runtime: Value,
}

impl MemfaultdConfig {
    pub fn load(config_path: &Path) -> eyre::Result<MemfaultdConfig> {
        let JsonConfigs {
            base: mut config_json,
            runtime,
        } = Self::parse_configs(config_path)?;
        Self::merge_into(&mut config_json, runtime);

        // Transform the JSON object into a typed structure.
        let config: MemfaultdConfig = serde_json::from_value(config_json)?;

        let mut validation_errors = vec![];
        if let Some(software_version) = &config.software_version {
            if let Err(e) = software_version_is_valid(software_version) {
                validation_errors.push(format!("  Invalid value for \"software_version\": {}", e));
            }
        }
        if let Some(software_type) = &config.software_type {
            if let Err(e) = software_type_is_valid(software_type) {
                validation_errors.push(format!("  Invalid value for \"software_type\": {}", e));
            }
        }

        match validation_errors.is_empty() {
            true => Ok(config),
            false => Err(eyre::eyre!("\n{}", validation_errors.join("\n"))),
        }
    }

    /// Parse config file from given path and returns (builtin+system config, runtime config).
    pub fn parse_configs(config_path: &Path) -> eyre::Result<JsonConfigs> {
        // Initialize with the builtin config file.
        let mut base: Value = Self::parse(include_str!("../../builtin.conf"))
            .wrap_err("Error parsing built-in configuration file")?;

        // Read and parse the user config file.
        let user_config = Self::parse(std::fs::read_to_string(config_path)?.as_str())
            .wrap_err(eyre!("Error reading {}", config_path.display()))?;

        // Merge the two JSON objects together
        Self::merge_into(&mut base, user_config);

        // Load the runtime config but only if the file exists. (Missing runtime config is not an error.)
        let runtime_config_path = Self::runtime_config_path_from_json(&base)?;
        let runtime = if runtime_config_path.exists() {
            Self::parse(fs::read_to_string(&runtime_config_path)?.as_str()).wrap_err(eyre!(
                "Error reading runtime configuration {}",
                runtime_config_path.display()
            ))?
        } else {
            Value::Object(serde_json::Map::new())
        };

        Ok(JsonConfigs { base, runtime })
    }

    /// Set and write boolean in runtime config.
    pub fn set_and_write_bool_to_runtime_config(&self, key: &str, value: bool) -> eyre::Result<()> {
        let config_string = match fs::read_to_string(self.runtime_config_path()) {
            Ok(config_string) => config_string,
            Err(e) => {
                if e.kind() == std::io::ErrorKind::NotFound {
                    "{}".to_string()
                } else {
                    return Err(eyre::eyre!("Failed to read runtime config: {}", e));
                }
            }
        };

        let mut config_val = Self::parse(&config_string)?;
        config_val[key] = Value::Bool(value);

        self.write_value_to_runtime_config(config_val)
    }

    /// Write config to runtime config file.
    ///
    /// This is used to write the config to a file that can be read by the memfaultd process.
    fn write_value_to_runtime_config(&self, value: Value) -> eyre::Result<()> {
        let runtime_config_path = self.runtime_config_path();
        fs::write(runtime_config_path, value.to_string())?;

        Ok(())
    }

    pub fn runtime_config_path(&self) -> PathBuf {
        PathBuf::from(self.persist_dir.clone()).join("runtime.conf")
    }

    // Parse a Memfaultd JSON configuration file (with optional C-style comments) and return a serde_json::Value object.
    fn parse(config_string: &str) -> eyre::Result<Value> {
        let json_text = string::remove_comments(config_string);
        let json: Value = serde_json::from_str(json_text.as_str())?;
        if !json.is_object() {
            return Err(eyre::eyre!("Configuration should be a JSON object."));
        }
        Ok(json)
    }

    /// Merge two JSON objects together. The values from the second one will override values in the first one.
    fn merge_into(dest: &mut Value, src: Value) {
        assert!(dest.is_object() && src.is_object());
        if let Value::Object(dest_map) = src {
            for (key, value) in dest_map {
                if let Some(obj) = dest.get_mut(&key) {
                    if obj.is_object() {
                        MemfaultdConfig::merge_into(obj, value);
                        continue;
                    }
                }
                dest[&key] = value;
            }
        }
    }

    pub fn generate_tmp_filename(&self, filename: &str) -> PathBuf {
        // Fall back to persist dir if tmp_dir is not set.
        let tmp_dir = self.tmp_dir.as_ref().unwrap_or(&self.persist_dir);
        PathBuf::from(tmp_dir.clone()).join(filename)
    }

    pub fn generate_persist_filename(&self, filename: &str) -> PathBuf {
        PathBuf::from(self.persist_dir.clone()).join(filename)
    }

    /// Generate the path to the runtime config file from a serde_json::Value object. This should include the "persist_dir" field.
    fn runtime_config_path_from_json(config: &Value) -> eyre::Result<PathBuf> {
        let mut persist_dir = PathBuf::from(
            config["persist_dir"]
                .as_str()
                .ok_or(eyre::eyre!("Config['persist_dir'] must be a string."))?,
        );
        persist_dir.push("runtime.conf");
        Ok(persist_dir)
    }
}

#[cfg(test)]
impl MemfaultdConfig {
    pub fn test_fixture() -> Self {
        use std::fs::write;
        use tempfile::tempdir;

        let tmp = tempdir().unwrap();
        let config_path = tmp.path().join("memfaultd.conf");
        write(&config_path, "{}").unwrap();
        MemfaultdConfig::load(&config_path).unwrap()
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_json_snapshot, with_settings};
    use rstest::rstest;

    use super::*;

    use crate::test_utils::set_snapshot_suffix;

    #[test]
    fn test_merge() {
        let mut c =
            serde_json::from_str(r#"{ "node": { "value": true, "valueB": false } }"#).unwrap();
        let j = serde_json::from_str(r#"{ "node2": "xxx" }"#).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r#"{"node":{"value":true,"valueB":false},"node2":"xxx"}"#
        );
    }

    #[test]
    fn test_merge_overwrite() {
        let mut c =
            serde_json::from_str(r#"{ "node": { "value": true, "valueB": false } }"#).unwrap();
        let j = serde_json::from_str(r#"{ "node": { "value": false }}"#).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r#"{"node":{"value":false,"valueB":false}}"#
        );
    }

    #[test]
    fn test_merge_overwrite_nested() {
        let mut c = serde_json::from_str(
            r#"{ "node": { "value": true, "valueB": false, "valueC": { "a": 1, "b": 2 } } }"#,
        )
        .unwrap();
        let j = serde_json::from_str(r#"{ "node": { "valueC": { "b": 42 } }}"#).unwrap();

        MemfaultdConfig::merge_into(&mut c, j);

        assert_eq!(
            serde_json::to_string(&c).unwrap(),
            r#"{"node":{"value":true,"valueB":false,"valueC":{"a":1,"b":42}}}"#
        );
    }

    #[rstest]
    #[case("empty_object")]
    #[case("with_partial_logs")]
    #[case("without_coredump_compression")]
    #[case("with_coredump_capture_strategy_threads")]
    #[case("with_log_to_metrics_rules")]
    #[case("with_connectivity_monitor")]
    #[case("with_sessions")]
    #[case("metrics_config")]
    fn can_parse_test_files(#[case] name: &str) {
        let input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/config/test-config")
            .join(name)
            .with_extension("json");
        // Verifies that the file is parsable
        let content = MemfaultdConfig::load(&input_path).unwrap();
        // And that the configuration generated is what we expect.
        // Use `cargo insta review` to quickly approve changes.
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(name, content);
        });
    }

    #[rstest]
    #[case("with_invalid_path")]
    #[case("with_invalid_swt_swv")]
    #[case("with_sessions_invalid_metric_name")]
    #[case("with_sessions_invalid_session_name")]
    fn will_reject_bad_config(#[case] name: &str) {
        let input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/config/test-config")
            .join(name)
            .with_extension("json");
        let result = MemfaultdConfig::load(&input_path);
        assert!(result.is_err());
    }

    #[rstest]
    #[case("no_file", None)]
    #[case("empty_object", Some("{}"))]
    #[case("other_key", Some(r#"{"key2":false}"#))]
    fn test_set_and_write_bool_to_runtime_config(
        #[case] test_name: &str,
        #[case] config_string: Option<&str>,
    ) {
        let mut config = MemfaultdConfig::test_fixture();
        let temp_data_dir = tempfile::tempdir().unwrap();
        config.persist_dir = AbsolutePath::try_from(temp_data_dir.path().to_path_buf()).unwrap();

        if let Some(config_string) = config_string {
            std::fs::write(config.runtime_config_path(), config_string).unwrap();
        }

        config
            .set_and_write_bool_to_runtime_config("key", true)
            .unwrap();

        let disk_config_string = std::fs::read_to_string(config.runtime_config_path()).unwrap();

        set_snapshot_suffix!("{}", test_name);
        insta::assert_json_snapshot!(disk_config_string);
    }
}
