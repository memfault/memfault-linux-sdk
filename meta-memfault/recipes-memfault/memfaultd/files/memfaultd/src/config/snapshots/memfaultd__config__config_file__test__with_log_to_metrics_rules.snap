---
source: memfaultd/src/config/config_file.rs
expression: content
---
{
  "persist_dir": "/media/memfault",
  "tmp_dir": null,
  "tmp_dir_min_headroom_kib": 10240,
  "tmp_dir_min_inodes": 100,
  "tmp_dir_max_usage_kib": 102400,
  "upload_interval_seconds": 3600,
  "heartbeat_interval_seconds": 3600,
  "enable_data_collection": false,
  "enable_dev_mode": false,
  "software_version": null,
  "software_type": null,
  "project_key": "",
  "base_url": "https://device.memfault.com",
  "swupdate": {
    "input_file": "/etc/swupdate.cfg",
    "output_file": "/tmp/swupdate.cfg"
  },
  "reboot": {
    "last_reboot_reason_file": "/media/last_reboot_reason",
    "capture_pstore": false
  },
  "coredump": {
    "compression": "gzip",
    "coredump_max_size_kib": 96000,
    "rate_limit_count": 5,
    "rate_limit_duration_seconds": 3600,
    "capture_strategy": {
      "type": "threads",
      "max_thread_size_kib": 32
    },
    "log_lines": 100
  },
  "fluent-bit": {
    "extra_fluentd_attributes": [],
    "bind_address": "127.0.0.1:5170",
    "max_buffered_lines": 1000,
    "max_connections": 4
  },
  "logs": {
    "rotate_size_kib": 10240,
    "rotate_after_seconds": 3600,
    "compression_level": 1,
    "max_lines_per_minute": 500,
    "log_to_metrics": {
      "rules": [
        {
          "type": "count_matching",
          "pattern": "ssh",
          "metric_name": "ssh_logins",
          "filter": {}
        },
        {
          "type": "count_matching",
          "pattern": "ssh",
          "metric_name": "ssh_logins",
          "filter": {
            "_SYSTEMD_UNIT": "sshd.service"
          }
        }
      ]
    },
    "storage": "persist",
    "source": "fluent-bit"
  },
  "mar": {
    "mar_file_max_size_kib": 10240,
    "mar_entry_max_age_seconds": 604800
  },
  "http_server": {
    "bind_address": "127.0.0.1:8787"
  },
  "battery_monitor": null,
  "connectivity_monitor": null,
  "sessions": null,
  "metrics": {
    "enable_daily_heartbeats": false,
    "system_metric_collection": {
      "enable": false,
      "poll_interval_seconds": 10,
      "processes": null,
      "disk_space": null,
      "network_interfaces": null
    },
    "statsd_server": null
  }
}
